Allow passing Mime\Part as attachment

tdutrion opened the pull request:
I ran into this problem when trying to use the mailer library in conjunction with Flysytem. My use case is: how to use a file from the filesystem (using the read method, so basically the file as a string) directly into the email, without writing it to the local disk (in memory only).

This addition allow to pass the mime part through the attachments array:

```
$pdfFile = new Mime\Part($this->filesystem->read('file.pdf'));
$pdfFile->setType('application/pdf')
             ->setId('file.pdf')
            ->setFileName('file.pdf')
            ->setEncoding(Mime\Mime::ENCODING_BASE64)
            ->setDisposition(Mime\Mime::DISPOSITION_ATTACHMENT);

$this->mailService->addAttachments([
    'file.pdf' => $pdfFile,
]);
```

Another thing this modification allows is to attach files inline rather than as a full attachment, which is useful to add inline images to an email.

Adding the ability to pass a callable in is useful in order to define an inline image from the configuration file for instance:

```
'attachments' => [
    'files' => [
        'logo.png' => function () {
            $file = new \Zend\Mime\Part(file_get_contents('logo.png'));
            $file->setType('image/png')
                ->setId('logo.png')
                ->setFileName('logo.png')
                ->setEncoding(\Zend\Mime\Mime::ENCODING_BASE64)
                ->setDisposition(\Zend\Mime\Mime::DISPOSITION_INLINE)
            ;
            return $file;
        },
    ],
],
```

This change is compatible with version 5 and 6, and I would be keen to see it on V5 as I am currently working with ZF2.4 LTS :)

acelaya commented:
I like the idea of allowing to pass a Mime object, but not the idea of Closures.
Defining Closures in config files is a bad practice, since it breaks configuration caching.
Maybe it would be good to allow to pass an array with all the properties of the Mime instance, instead of just passing the name.
That would allow to preconfigure the Mime without breaking config caching.

Another thing that would be interesting is allowing to pass files as streams (resources). If I'm correct, flysystem allows to get a file stream.
That way you don't need to completely read big files in memory.

Do you want to implement any of this?
By the way, the build is broken.

tdutrion commented:
I added #141 with regards to travisci not passing the tests, I don't know how to fix it.

Regarding the Mime Part factory, none exist at the moment (I guess), and I couldn't bother create one for now. I added a way to use a classname rather than a closure (invokable class obviously) so that works with opcache. Unfortunately, I couldn't link it to the service manager, but that is something we may want to look into, so that the invokable class could have dependencies and therefore be a little bit more useful.

FlySystem does allow streams, I never tried it but saw the documentation. I can look into that after the end of year holidays that are a bit busy!

acelaya commented:
There is no need to explicitly crate a factory, we just need to allow the `MailServiceInterface::addAttachment` method to allow an array to be passed.

That way, we could do things like this:

```php
$mailService->addAttachment([
    'path' => 'data/attachments/logo.png',
    'type' => 'image/png',
    'id' => 'logo.png',
    'file_name' => 'logo.png',
    'encoding' => \Zend\Mime\Mime::ENCODING_BASE64,
    'disposition' => \Zend\Mime\Mime::DISPOSITION_INLINE,
    // ... Not sure if there's anything else that could be relevant
]);
```

Or multiple files at a time
```php
$mailService->addAttachments([
    'foo' => [
        'path' => 'data/attachments/logo.png',
        'type' => 'image/png',
        'id' => 'logo.png',
        'file_name' => 'logo.png',
        'encoding' => \Zend\Mime\Mime::ENCODING_BASE64,
        'disposition' => \Zend\Mime\Mime::DISPOSITION_INLINE,
    ],
    'bar' => [
        // ...
    ]
]);
```

Or via configuration

```php
'attachments' => [
     'files' => [
        'logo.png' => [
            'path' => 'data/attachments/logo.png',
            'type' => 'image/png',
            'id' => 'logo.png',
            'file_name' => 'logo.png',
            'encoding' => \Zend\Mime\Mime::ENCODING_BASE64,
            'disposition' => \Zend\Mime\Mime::DISPOSITION_INLINE,
        ],
    ],
],
```

We would only need to update the protected `attachFiles` method take into account that attachments could be passed in the form of arrays, casting them to a Mime\Part object.

What do you think about this approach.
We could start with this and then focus on accepting streams too in another pull request.

acelaya commented:
Since I don't see much progress, I'll implement it as we discussed.
I'll support adding attachments as resources, as arrays and as `Zend\Mime\Part` objects (I'll try to make sure it works with Flysystem returned streams, but at least the array or `Zend\Mime\Part` approaches should work)

acelaya commented:
I'll try to backport it to v5, by the way.

acelaya commented:
It's already merged in v6.4.0 and 5.3.0.
Now, it is possible to attach files as string file path, as array, as resource and as `Zend\Mime\Part`.

I hope it helps :)
